### ISSUE TITLE
1/I is automatically simplified to -I

### ISSUE THREAD

**Issue by proy87** [sympy/sympy#19706]
Should this autosimplification be performed?

**Comment by gschintgen** [comment]
Yes, why should it not? It's a simple and straightforward computation, just as e.g. `2*3` which is also auto-simplified.

**Comment by smichr** [comment]
imaginary denominators are made read at instantiation:
```python
>>> 1/(1+I)
(1 - I)/2
>>> 1/((I+2)*(1+I))
(1 - I)*(2 - I)/10
```

**Comment by oscarbenjamin** [comment]
Personally I would prefer it if this simplification was not performed automatically but I think that about most automatic simplification. This one bothers me as well:
```julia
In [1]: 1/sqrt(2)                                                                                                                              
Out[1]: 
√2
──
2 
```

**Comment by friyaz** [comment]
I agree with Oscar. Maybe we can have some external flag that disables such automatic simplification.

**Comment by smichr** [comment]
maybe a `proper_fraction` flag?

**Comment by oscarbenjamin** [comment]
I would rather these just didn't happen unless some sort of function was called like ratsimp, simplify, cancel, maybe expand etc.

**Comment by jksuom** [comment]
How about comparisons like `1/I == -I`?
```
In [3]: Pow(I, -1, evaluate=False) == -I
Out[3]: False
```

**Comment by oscarbenjamin** [comment]
Some comparisons wouldn't be able to resolve without some kind of normalisation/simplification. That's already the case in many other situations:
```julia
In [3]: x*(y + z) == x*y + x*z                                                                                                                 
Out[3]: False
```
The problem with automatic simplification is that it makes it impossible to express the result in some ways e.g.:
```julia
In [4]: sin(k).rewrite(exp)                                                                                                                    
Out[4]: 
   ⎛ ⅈ⋅k    -ⅈ⋅k⎞ 
-ⅈ⋅⎝ℯ    - ℯ    ⎠ 
──────────────────
        2 

In [5]: (exp(I*k) - exp(-I*k)) / (2*I)                                                                                                         
Out[5]: 
   ⎛ ⅈ⋅k    -ⅈ⋅k⎞ 
-ⅈ⋅⎝ℯ    - ℯ    ⎠ 
──────────────────
        2  
```
I've never seen anyone write the formula like that but there's no way to get the `I` on the bottom without `evaluate=False` and `evaluate=False` doesn't (can't) work in general.

We would need to create light-weight functions for performing this kind of normalisation of powers of `I`, `sqrt(2)` etc. if it didn't happen by default.

The difficulty is all the code that is depending on this happening automatically as it currently does. Similar problems would apply to removing the 2-arg Rational/Add Mul.

**Comment by proy87** [comment]
@gschintgen , `2**3` is an operation that is evaluated. `1/I` can't be calculated in such manner (we need to multiply both numerator and denominator by `I` and use the fact that `i**2=-1`). Why `1/(I+2)**2` is not simplified then? Same for `1/sqrt(2)`.
As  far as I know `sin(x)**2 + cos(x)**2` is not autosimplified into `1`, although it is a one-step identity, while `1/I=I/I**2=I/-1=-I` is two-step.
To simplify `1/(I+2)`, one need to perform several operations and in many cases it is desirable to leave it unchanged.

**Comment by gschintgen** [comment]
> @gschintgen , `2**3` is an operation that is evaluated. `1/I` can't be calculated in such manner (we need to multiply both numerator and denominator by `I` and use the fact that `i**2=-1`). Why `1/(I+2)**2` is not simplified then? Same for `1/sqrt(2)`.

First of all: this is my very personal opinion as a user that mostly uses SymPy in interactive use and I don't pretend to speak for the SymPy project as a whole. For the interactive usecase I expect a CAS to do plain numeric computations without causing too much hassle. I respectfully disagree that `1/I` is a multi-step operation: I see it as a single division in the default field of complex numbers.  IMO addition, multiplication, subtraction, division of plain numeric arguments should auto-evaluate as far as reasonable and that definitely includes basic computations (with e.g. rational real/imaginary parts) in C. Obviously the crux is the part "as far as reasonable" (think `2**100**100` etc.)

As far as I'm concerned, I'd prefer if e.g. `(2+I)*5` were auto-evaluated just as `(S(2)+3)*5` is.

I do understand the opposite point of view though, as expressed by @oscarbenjamin. But it seems to me that a default of no evaluation is more suited for library usage or development of library code where such "automagic" changes can be a nuisance and may also introduce performance issues. I guess it would be best to have one or more global flags to control auto-evaluation but that will be a herculean task to implement.

**Comment by proy87** [comment]
@gschintgen , you are right. For different cases, one needs different output. A flag would be good.

**Comment by oscarbenjamin** [comment]
I think that it is possible to distinguish between the library use and the interactive user usage. The way that matrix expressions work does this quite nicely I think. The key is that you get a different result from `MatMul` then you would from `*` like this:
```julia
In [10]: M = MatrixSymbol('M', 2, 2)                                                                                                           

In [11]: MatMul(M, M)                                                                                                                          
Out[11]: M⋅M

In [12]: MatMul(M, M).doit()                                                                                                                   
Out[12]: 
 2
M 

In [13]: M * M                                                                                                                                 
Out[13]: 
 2
M 
```
This means that the user can get what they want using `*` but the `MatMul` basic subclass itself does not automatically evaluate. This is important because it means that we can represent the "unevaluated" result without braking the func-args invariant:
```
     obj  ==  obj.func(*obj.args)   [ for all Basic instances obj ]
```
By contrast with `Pow` we get automatic evaluation e.g.
```julia
In [14]: Pow(I, -1)                                                                                                                            
Out[14]: -ⅈ

In [15]: I**-1                                                                                                                                 
Out[15]: -ⅈ

In [16]: 1/I                                                                                                                                   
Out[16]: -ⅈ
```
The only way to get the unevaluated result is to use `evaluate=False` but that breaks the func-args invariant:
```julia
In [18]: obj = Pow(I, -1, evaluate=False)                                                                                                      

In [19]: obj                                                                                                                                   
Out[19]: 
1
─
ⅈ

In [20]: obj.doit()                                                                                                                            
Out[20]: -ⅈ

In [21]: obj.func(*obj.args)                                                                                                                   
Out[21]: -ⅈ

In [22]: obj.func(*obj.args) == obj                                                                                                            
Out[22]: False
```
The problem with breaking the func-args invariant is that other `Basic` operations like `subs` or `Add/Mul.flatten()` will trigger the evaluation:
```julia
In [25]: obj                                                                                                                                   
Out[25]: 
1
─
ⅈ

In [26]: x*obj                                                                                                                                 
Out[26]: -ⅈ⋅x

In [27]: expr = Mul(x, obj, evaluate=False)                                                                                                    

In [28]: expr                                                                                                                                  
Out[28]: 
x
─
ⅈ

In [29]: expr.subs(x, y)                                                                                                                       
Out[29]: -ⅈ⋅y
```
It is possible as in the case of matrix expressions to have `Pow(x, y)` and `x**y` do different things and also `Pow(x, -1)` and `1/x` do not need to be equivalent. I am primarily concerned with the kind of automatic evaluation that comes from using `Pow` directly because it makes many manipulations impossible.

Many users though would prefer not to have this kind of "evaluation" when using the operators like `/` and `**` and I generally find myself agreeing with them as well. I personally don't think that `sqrt(2)/2` is simpler/better than `1/sqrt(2)` although I understand that it is done to get a canonical representation. I see also that many University students seem to prefer to write things as `sqrt(2)/2` which I think is because that's what some pocket calculators do.

> As far as I'm concerned, I'd prefer if e.g. `(2+I)*5` were auto-evaluated just as `(S(2)+3)*5` is.

Actually this is "evaluated" but it's a whole separate can of auto-evaluation worms:
```julia
In [30]: (2+I)*5                                                                                                                               
Out[30]: 10 + 5⋅ⅈ
```
This example works because of the following rule that is implemented in `Mul.flatten`:
https://github.com/sympy/sympy/blob/b1078cd096aa891ebadf5d941ab83268f8b03a80/sympy/core/mul.py#L193
This is what I referred to above as the 2-arg Mul. The rule says that for `Mul(expr1, expr2)` (only 2 args to the Mul) if the one arg is Rational and the other is an Add then the rational should be distributed over the Add. That's why `(2+I)*5` distributes because `5` is Rational and `2+ I` is an `Add`. There are serious problems with this though:

1. The special-case nature of this is obvious if you look at this branch compared to the rest of the `Mul.flatten` routine. If you looked at what the rest of the function does I don't think you'd look at that and then decide to add this at the beginning. I imagine that it predates the rest of the routine reaching its current form.

2. This rule breaks associativity:
```
In [36]: (a*b)*c                                                                                                                               
Out[36]: x⋅(10 + 5⋅ⅈ)

In [37]: a*(b*c)                                                                                                                               
Out[37]: 5⋅x⋅(2 + ⅈ)

In [38]: (a*b)*c == a*(b*c)                                                                                                                    
Out[38]: False
```

3. It breaks commutativity (when applicable):
```julia
In [66]: a*b*c                                                                                                                                 
Out[66]: x⋅(10 + 5⋅ⅈ)

In [67]: a*c*b                                                                                                                                 
Out[67]: 5⋅x⋅(2 + ⅈ)

In [68]: a*b*c == a*c*b                                                                                                                        
Out[68]: False
```

4. Breaking associativity/commutativity leads to problems because of other code that assumes these properties. The most common reason that users/contributors hit up against the 2-arg Mul as that you can't copy paste `str(obj)` because the printing code has made the incorrect (but not unreasonable) assumption of associativity:
```julia
In [54]: obj = Mul(2, (x+y), 1+I)                                                                                                              

In [55]: obj                                                                                                                                   
Out[55]: 2⋅(1 + ⅈ)⋅(x + y)

In [56]: str(obj)                                                                                                                              
Out[56]: '2*(1 + I)*(x + y)'

In [57]: 2*(1 + I)*(x + y)                                                                                                                     
Out[57]: (2 + 2⋅ⅈ)⋅(x + y)

In [58]: obj == eval(str(obj))                                                                                                                 
Out[58]: False
```

5. The 2-arg Mul special cases concrete numbers over symbols which leads to code that works for numeric values but not for symbolic values. This is because there is no automatic distribution if we don't have a Rational:
```julia
In [59]: x*(y + z)                                                                                                                             
Out[59]: x⋅(y + z)

In [70]: 2*(x + y)                                                                                                                             
Out[70]: 2⋅x + 2⋅y
```
It's clear that we don't want such expressions to always expand. We need some sort of function that is specifically for the purpose of distributing one expression over an Add and then that should be used where it is needed rather than seeming to work in the case of rationals but failing in the symbolic case. Not only do I think that `Mul(expr1, expr2)` should not do this but I also think that `Mul(expr1, expr2).doit()` and `expr1*expr2` should not do this either. There should be a function like `distribute` or something.

**Comment by gschintgen** [comment]
> > As far as I'm concerned, I'd prefer if e.g. `(2+I)*5` were auto-evaluated just as `(S(2)+3)*5` is.
> 
> Actually this is "evaluated" [...]

oops, you're right of course. When typing my comment I "simplified" the example and messed it up. What I meant was e.g. `(2 + I)*(5 - I)` which is a single multiplication of two (complex) numbers (at least from a mathematical point of view, not SymPy internals).

Thanks anyway for the very detailed comment, @oscarbenjamin. I guess I wouldn't mind having more automatic evaluation in the numeric case, but I can see how such a discrepancy between numeric and symbolic cases can quickly lead to flaky library code... 

The matrix examples are nice indeed. The noun/verb distinction in other parts is also interesting in this context (integrate vs. Integral). I guess Set.intersect/Intersection() could/should also be fixed to properly distinguish between both cases.

**Comment by oscarbenjamin** [comment]
> I guess Set.intersect/Intersection() could/should also be fixed to properly distinguish between both cases.

Yes, I think so. See also #17372 and #17280. During #17372 I hit on so many problems in sets that I ended up rewriting lots of other stuff without achieving the original intention.

**Comment by asmeurer** [comment]
I agree `1/I` should not evaluate. It's quite common to have an `i` in the denominator of an equation. For example, Cauchy's integral formula is almost always specified using `1/(2 pi i)`, not `-i/(2 pi)`. If you work enough with complex numbers you can easily "see" `1/i` and `-i` as the same thing, but we shouldn't require people to be so practiced. 

Some of the other evaluation cases mentioned here like rationalization of numbers should be considered separately, IMO. The only one that is relevant is `I**2`, which I personally think should continue to automatically evaluate, along with other positive integer powers of `I`. 

It's worth mentioning my general views of automatic evaluation that I put on the wiki some time ago: https://github.com/sympy/sympy/wiki/Automatic-Simplification. `1/I` -> `-I` clearly fails rule 4, "the rule is so trivial no on would never want it done", based on what people here have said. `I**2` -> `=1` on the other hand tends to be something that almost everyone would expect to automatically evaluate.

**Comment by asmeurer** [comment]
> imaginary denominators are made read at instantiation:

I didn't realize it did this even for any complex number. Even if 1/I evaluates I don't think 1 + I should.

Also I'm -1 to any kind of flag here. Either it should evaluate or it shouldn't. If there is a desire for a flag that is a sign that we just shouldn't evaluate it and have some functions that can evaluate it post hoc if desired.

**Comment by friyaz** [comment]
> If there is a desire for a flag that is a sign that we just shouldn't evaluate it and have some functions that can evaluate it post hoc if desired.

Changing the default behavior will break many tests and It may be very difficult to fix some of them.
